checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -0,0 +1,164 @@
+
+  /** The control flow graph. */
+  private ControlFlowGraph cfg;
+
+  /**
+   * Creates a MustCallInferenceLogic. If the whole program inference is not null, the type
+   * factory's postAnalyze method would instantiate a new MustCallInferenceLogic using this
+   * constructor and then call {@link #runInference()}.
+   *
+   * @param typeFactory the type factory
+   * @param cfg the ControlFlowGraph
+   */
+  MustCallInferenceLogic(ResourceLeakAnnotatedTypeFactory typeFactory, ControlFlowGraph cfg) {
+    this.typeFactory = typeFactory;
+    this.cfg = cfg;
+    OWNING = AnnotationBuilder.fromClass(this.typeFactory.getElementUtils(), Owning.class);
+  }
+
+  /**
+   * It checks method invocations for fields with non-empty @MustCall obligation along all paths to
+   * the regular exit point in the method body.
+   */
+  void runInference() {
+    Set<Block> visited = new HashSet<>();
+    Deque<Block> worklist = new ArrayDeque<>();
+    Block entry = this.cfg.getEntryBlock();
+
+    worklist.add(entry);
+    visited.add(entry);
+
+    while (!worklist.isEmpty()) {
+      Block current = worklist.remove();
+
+      for (Node node : current.getNodes()) {
+        if (node instanceof MethodInvocationNode) {
+          checkForMustCallInvocationOnField((MethodInvocationNode) node);
+        }
+      }
+
+      propagateRegPaths(current, visited, worklist);
+    }
+  }
+
+  /**
+   * If the receiver of {@code mNode} is a possible owning field and the method invocation satisfies
+   * the field's must call obligation, then adds owning annotation for that field.
+   *
+   * @param mNode the MethodInvocationNode
+   */
+  private void checkForMustCallInvocationOnField(MethodInvocationNode mNode) {
+    Node receiver = mNode.getTarget().getReceiver();
+    if (receiver.getTree() == null) {
+      return;
+    }
+
+    Element receiverEl = TreeUtils.elementFromTree(receiver.getTree());
+
+    if (typeFactory.isPossibleOwningField(receiverEl)) {
+      Element method = TreeUtils.elementFromTree(mNode.getTree());
+      List<String> mustCallValues = typeFactory.getMustCallValue(receiverEl);
+
+      // Because we assumed that any must call annotation has at most one method, the following
+      // check is enough to decide whether the receiver is an owning field
+      if (mustCallValues.size() == 1
+          && mustCallValues.contains(method.getSimpleName().toString())) {
+        owningFields.add(receiverEl);
+      }
+    }
+  }
+
+  /**
+   * updates worklist with the next block along all paths to the regular exit point.
+   *
+   * @param curBlock the current block
+   * @param visited set of blocks already on the worklist
+   * @param worklist current worklist
+   */
+  private void propagateRegPaths(Block curBlock, Set<Block> visited, Deque<Block> worklist) {
+
+    List<Block> successors = getSuccessors(curBlock);
+
+    for (Block b : successors) {
+      if (b.getType() == Block.BlockType.SPECIAL_BLOCK) {
```

Comment:
Add a comment that the block must be the regular exit, since we do not propagate to exceptional successors